test/integration/serverless/log_normalize.py

```diff
@@ -0,0 +1,148 @@
+            replace(r'\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z', 'TIMESTAMP'),
+            replace(r'\d{4}\/\d{2}\/\d{2}\s\d{2}:\d{2}:\d{2}', 'TIMESTAMP'),
+            replace(r'\d{4}-\d{2}-\d{2}\s\d{2}:\d{2}:\d{2}', 'TIMESTAMP'),
+            replace(r'([a-zA-Z0-9]{8}-[a-zA-Z0-9]{4}-[a-zA-Z0-9]{4}-[a-zA-Z0-9]{4}-[a-zA-Z0-9]{12})', r'XXX'),
+            replace(stage, 'XXXXXX'),
+            replace(r'(architecture:)(x86_64|arm64)', r'\1XXX'),
+            sort_by(lambda log: log["message"]["message"]),
+            # ignore a Lambda error that occurs sporadically for log-csharp see here for more info:
+            # https://repost.aws/questions/QUq2OfIFUNTCyCKsChfJLr5w/lambda-function-working-locally-but-crashing-on-aws
+            # TODO
+            # perl -n -e "print unless /LAMBDA_RUNTIME Failed to get next invocation. No Response from endpoint/ or \
+            #  /An error occurred while attempting to execute your code.: LambdaException/ or \
+            #  /terminate called after throwing an instance of 'std::logic_error'/ or \
+            #  /basic_string::_M_construct null not valid/" |
+    ]
+
+def normalize_traces(stage):
+    return [
+            require(r'BEGINTRACE.*ENDTRACE'),
+            exclude(r'BEGINTRACE'),
+            exclude(r'ENDTRACE'),
+            replace(r'(ts":)[0-9]{10}', r'\1XXX'),
+            replace(r'((startTime|endTime|traceID|trace_id|span_id|parent_id|start|system.pid)":)[0-9]+', r'\1null'),
+            replace(r'((tracer_version|language_version)":)["a-zA-Z0-9~\-\.\_]+', r'\1null'),
+            replace(r'(duration":)[0-9]+', r'\1null'),
+            replace(r'((datadog_lambda|dd_trace)":")[0-9]+\.[0-9]+\.[0-9]+', r'\1X.X.X'),
+            replace(r'(,"request_id":")[a-zA-Z0-9\-,]+"', r'\1null"'),
+            replace(r'(,"runtime-id":")[a-zA-Z0-9\-,]+"', r'\1null"'),
+            replace(r'(,"system.pid":")[a-zA-Z0-9\-,]+"', r'\1null"'),
+            replace(r'("_dd.no_p_sr":)[0-9\.]+', r'\1null'),
+            replace(r'("architecture":)"(x86_64|arm64)"', r'\1"XXX"'),
+            replace(r'("process_id":)[0-9]+', r'\1null'),
+            replace(stage, 'XXXXXX'),
+            exclude(r'[ ]$'),
+    ]
+
+#####################
+# BEGIN NORMALIZERS #
+#####################
+
+def replace(pattern, repl):
+    comp = re.compile(pattern, flags=re.DOTALL)
+    def _replace(log):
+        return comp.sub(repl, log)
+    return _replace
+
+def exclude(pattern):
+    comp = re.compile(pattern, flags=re.DOTALL)
+    return replace(pattern, '')
+
+def require(pattern):
+    comp = re.compile(pattern, flags=re.DOTALL)
+    def _require(log):
+        match = comp.search(log)
+        if not match:
+            return ''
+        return match.group(0)
+    return _require
+
+def sort_by(key):
+    def _sort(log):
+        log_json = json.loads(log, strict=False)
+        log_sorted = sorted(log_json, key=key)
+        return json.dumps(log_sorted)
+    return _sort
```

Comment:
This normalizer is where the magic is.

You'll remember that most of the failures we receive in our integration tests are just logs coming in out of order. To address this, I've created a sort normalizer which will sort the logs by their messages, instead of their timestamps.

I believe this change in sort order is safe. The backend will always sort logs in their correct order based on their timestamp. So, we here do not actually need to worry about their sort order.